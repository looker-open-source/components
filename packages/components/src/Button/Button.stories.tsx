/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { AddCircle, Delete } from '@styled-icons/material'
import type { Page } from 'puppeteer'
import React from 'react'
import type { Story } from '@storybook/react/types-6-0'
import { defaultArgTypes as argTypes } from '../../../../apps/storybook/src/defaultArgTypes'
import { Button } from './Button'
import type { ButtonProps } from './types'

const Template: Story<ButtonProps> = args => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Focused = Template.bind({})
Focused.args = {
  ...Basic.args,
  margin: 'small',
}

const beforeScreenshot = async (page: Page) => {
  const button = await page.$('button')
  await button?.type(' ')
  await page.waitForTimeout(50)
}
Focused.parameters = {
  beforeScreenshot,
  docs: { disable: true },
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const CriticalFocused = Template.bind({})
CriticalFocused.args = {
  ...Critical.args,
  margin: 'small',
}

CriticalFocused.parameters = {
  beforeScreenshot,
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const NeutralFocused = Template.bind({})
NeutralFocused.args = {
  ...Neutral.args,
  margin: 'small',
}

NeutralFocused.parameters = {
  beforeScreenshot,
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: <AddCircle />,
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: <Delete />,
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  argTypes,
  component: Button,
  title: 'Button',
}
