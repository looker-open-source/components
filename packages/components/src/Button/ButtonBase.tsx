/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  radii,
  RadiusSizes,
  reset,
  SemanticColors,
  space,
  SpaceProps,
} from '@looker/design-tokens'
import { rgba } from 'polished'
import React, { forwardRef, Ref, useState } from 'react'
import styled, { css } from 'styled-components'
import {
  minWidth,
  MinWidthProps,
  maxWidth,
  MaxWidthProps,
  WidthProps,
  width,
} from 'styled-system'
import { buttonSize, ButtonSizes, ButtonSizeProps } from './size'
import { ButtonIcon, buttonIcon, ButtonIconProps } from './icon'

export interface CustomizableButtonAttributes {
  borderRadius: RadiusSizes
}

export const CustomizableButtonAttributes: CustomizableButtonAttributes = {
  borderRadius: 'medium',
}

type ButtonColors = keyof SemanticColors

export interface ButtonBaseProps
  extends Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'>,
    ButtonSizeProps,
    MaxWidthProps,
    MinWidthProps,
    WidthProps,
    SpaceProps {
  type?: 'button' | 'submit' | 'reset'

  /**
   * Used along with `href` for an anchor link styled as a Button
   */
  forwardedAs?: 'a'

  /**
   * Defines the color of the button. Can be the string name of a color listed in the color theme, or a color object.
   * @default "primary"
   */
  color?: ButtonColors

  focusVisible?: boolean
}

export interface ButtonProps extends ButtonBaseProps, ButtonIconProps {
  size?: ButtonSizes
  /**
   * If true, the button's width will be set to 100%.
   */
  fullWidth?: boolean
}

export const buttonCSS = css<ButtonBaseProps>`
  ${reset}
  ${maxWidth}
  ${minWidth}
  ${width}

  ${props =>
    props.focusVisible &&
    `
    box-shadow: 0 0 0 0.15rem
      ${rgba(
        props.theme.colors.semanticColors[props.color || 'primary'].main,
        0.25
      )};
  `}

  align-items: center;
  border-radius: ${radii[CustomizableButtonAttributes.borderRadius]};
  cursor: pointer;
  display: inline-flex;
  font-weight: 600;
  justify-content: center;
  line-height: 1;
  outline: none;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;

  &[disabled] {
    cursor: default;
    filter: grayscale(0.3);
    opacity: 0.25;
  }

  ${buttonSize}
  ${space}
`

const ButtonOuter = styled.button<ButtonProps>`
  ${buttonCSS}
  ${props => props.fullWidth && `width: 100%;`}
`

const ButtonJSX = forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      children,
      iconBefore,
      iconAfter,
      onBlur,
      onKeyDown,
      ...restProps
    } = props

    const [isFocusVisible, setFocusVisible] = useState(false)

    const handleOnKeyUp = (event: React.KeyboardEvent<HTMLButtonElement>) => {
      setFocusVisible(true)
      onKeyDown && onKeyDown(event)
    }

    const handleOnBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
      setFocusVisible(false)
      onBlur && onBlur(event)
    }

    return (
      <ButtonOuter
        focusVisible={isFocusVisible}
        onKeyUp={handleOnKeyUp}
        onBlur={handleOnBlur}
        {...restProps}
        ref={ref}
      >
        {iconBefore && <ButtonIcon name={iconBefore} />}
        {children}
        {iconAfter && <ButtonIcon name={iconAfter} />}
      </ButtonOuter>
    )
  }
)

ButtonJSX.displayName = 'ButtonJSX'

export const ButtonBase = styled(ButtonJSX)<ButtonProps>`
  ${buttonIcon}
`

ButtonBase.defaultProps = { size: 'medium' }
