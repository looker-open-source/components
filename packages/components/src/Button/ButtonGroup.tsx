/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import uniqueId from 'lodash/uniqueId'
import xor from 'lodash/xor'
import React, { ChangeEvent, FC, forwardRef, Ref } from 'react'
import styled from 'styled-components'
import { useControlWarn } from '../utils'
import { ButtonItem, ButtonItemLabel } from './ButtonItem'
import { ButtonGroupOrToggleProps, ButtonSet } from './ButtonSet'

const ButtonGroupFactory = forwardRef(
  (
    { onChange, value, ...props }: ButtonGroupOrToggleProps<string[]>,
    ref: Ref<HTMLDivElement>
  ) => {
    const isControlled = useControlWarn({
      controllingProps: ['onChange', 'value'],
      isControlledCheck: () => onChange !== undefined,
      name: 'ButtonGroup',
    })

    function handleChange(e: ChangeEvent<HTMLInputElement>) {
      const newValue = xor(value, [e.target.value])
      onChange && onChange(newValue)
    }

    return (
      <ButtonSet
        {...props}
        ref={ref}
        {...(isControlled
          ? {
              onChange: handleChange,
              value,
            }
          : { name: props.name || uniqueId() })}
      />
    )
  }
)

export const ButtonGroup = styled<FC<ButtonGroupOrToggleProps<string[]>>>(
  ButtonGroupFactory
)`
  ${ButtonItem} {
    height: 36px;
    border-radius: 4px;

    + ${ButtonItem} {
      margin-left: ${(props) => props.theme.space.xxsmall};
    }
  }

  ${ButtonItemLabel} {
    border-style: solid;
    border-width: 1px;
  }
`
