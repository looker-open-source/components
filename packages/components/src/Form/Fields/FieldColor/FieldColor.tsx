/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, {
  useState,
  ChangeEvent,
  FormEvent,
  forwardRef,
  Ref,
  useEffect,
} from 'react'
import styled from 'styled-components'
import { useID, useWrapEvent } from '../../../utils'
import { usePopover, PopoverContent } from '../../../Popover'
import { InputText, InputTextProps } from '../../Inputs/InputText'
import { Field, FieldProps, omitFieldProps, pickFieldProps } from '../Field'
import { FormControl } from '../../FormControl'
import { useFormContext } from '../../Form'
import {
  HueSaturation,
  polarbrightness2hsv,
  SimpleHSV,
  white,
} from './ColorWheel/color_wheel_utils'
import { ColorWheel } from './ColorWheel'
import { LuminositySlider } from './LuminositySlider'
import { Swatch } from './Swatch'
import {
  hsv2hex,
  simpleHSVtoFormattedColorString,
  str2simpleHsv,
} from './utils/color_format_utils'
import { isValidColor } from './utils/color_utils'

const colorWheelSize = 164

export interface FieldColorProps
  extends FieldProps,
    Omit<InputTextProps, 'height' | 'size'> {
  /**
   * If true, hides input and only show color swatch.
   */
  hideInput?: boolean
  value?: string
  defaultValue?: string
}

const createEventWithHSVValue = (
  color: SimpleHSV | string
): ChangeEvent<HTMLInputElement> => {
  return {
    currentTarget: {
      value:
        typeof color === 'string'
          ? color
          : simpleHSVtoFormattedColorString(color),
    },
    target: {
      value:
        typeof color === 'string'
          ? color
          : simpleHSVtoFormattedColorString(color),
    },
  } as ChangeEvent<HTMLInputElement>
}

function getColorFromText(text?: string) {
  const initialWhite = polarbrightness2hsv(white())
  return text && isValidColor(text) ? str2simpleHsv(text) : initialWhite
}

export const FieldColorComponent = forwardRef(
  (
    {
      hideInput,
      id,
      onChange,
      onFocus,
      onBlur,
      value,
      defaultValue,
      ...props
    }: FieldColorProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const inputID = useID(id)
    const validationMessage = useFormContext(props)
    const initialWhite = polarbrightness2hsv(white())
    const initialColor = getColorFromText(value || defaultValue)

    const [color, setColor] = useState<SimpleHSV>(initialColor)
    const [inputTextValue, setInputTextValue] = useState(
      value || defaultValue || ''
    )
    const [isFocused, setIsFocused] = useState(false)

    const handleFocus = () => setIsFocused(true)
    const handleBlur = () => setIsFocused(false)
    const wrappedOnFocus = useWrapEvent(handleFocus, onFocus)
    const wrappedOnBlur = useWrapEvent(handleBlur, onBlur)

    useEffect(() => {
      if (value && value !== inputTextValue) {
        setColor(str2simpleHsv(value))
        !isFocused && setInputTextValue(value)
      }
    }, [isFocused, value, inputTextValue])

    const callOnChange = (newColor: SimpleHSV | string) => {
      if (!onChange || !newColor) return
      onChange(createEventWithHSVValue(newColor))
    }

    const setColorState = (newColor: SimpleHSV) => {
      setColor(newColor)
      newColor && setInputTextValue(simpleHSVtoFormattedColorString(newColor))
      callOnChange(newColor)
    }

    const handleColorChange = (hs: HueSaturation) =>
      setColorState({ ...hs, v: color.v })

    const handleSliderChange = (event: FormEvent<HTMLInputElement>) =>
      setColorState({
        ...color,
        v: Number(event.currentTarget.value) / 100,
      })

    const handleInputTextChange = (event: FormEvent<HTMLInputElement>) => {
      const newValue = event.currentTarget.value
      setInputTextValue(newValue)

      const isValid = isValidColor(newValue)
      callOnChange(isValid ? newValue : initialWhite)
      setColor(getColorFromText(event.currentTarget.value))
    }

    const content = (
      <PopoverContent display="flex" flexDirection="column">
        <ColorWheel
          size={colorWheelSize}
          hue={color.h}
          saturation={color.s}
          value={color.v}
          onColorChange={handleColorChange}
        />
        <LuminositySlider
          min={0}
          max={100}
          step={1}
          value={color.v * 100}
          width={colorWheelSize}
          onChange={handleSliderChange}
        />
      </PopoverContent>
    )

    const { open, popover, ref: triggerRef } = usePopover({ content })

    return (
      <Field
        {...pickFieldProps(props)}
        id={inputID}
        validationMessage={validationMessage}
      >
        <FormControl alignLabel="left">
          <Swatch
            ref={triggerRef}
            color={hsv2hex(color)}
            borderRadius={hideInput ? 'medium' : 'none'}
            borderTopLeftRadius="medium"
            borderBottomLeftRadius="medium"
            border="1px solid"
            borderRight={hideInput ? undefined : 'none'}
            disabled={props.disabled}
            onClick={open}
          />
          {!props.disabled && popover}
          {!hideInput && (
            <InputText
              {...omitFieldProps(props)}
              aria-describedby={`${id}-describedby`}
              id={inputID}
              ref={ref}
              borderRadius="none"
              borderTopRightRadius="medium"
              borderBottomRightRadius="medium"
              validationType={validationMessage && validationMessage.type}
              onChange={handleInputTextChange}
              value={inputTextValue}
              onFocus={wrappedOnFocus}
              onBlur={wrappedOnBlur}
            />
          )}
        </FormControl>
      </Field>
    )
  }
)

FieldColorComponent.displayName = 'FieldColorComponent'

export const FieldColor = styled(FieldColorComponent)``
