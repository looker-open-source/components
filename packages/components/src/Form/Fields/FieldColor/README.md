# FieldColor

```tsx
function Basic(props: FieldColorProps) {
  const [value, setValue] = useState('purple');
  const handleChange = (e: ChangeEvent<HTMLInputElement>) =>
    setValue(e.target.value);

  return (
    <FieldColor
      {...props}
      label="Basic"
      value={value}
      onChange={handleChange}
    />
  );
}
```
