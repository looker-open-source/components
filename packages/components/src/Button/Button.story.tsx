/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { Button } from './Button'
import { ButtonProps } from './types'

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: 'CircleAdd',
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: 'Trash',
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  component: Button,
  title: 'Button',
}
