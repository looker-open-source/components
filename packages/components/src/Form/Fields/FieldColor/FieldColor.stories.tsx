/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import type { Story } from '@storybook/react/types-6-0'
import type { FormEvent } from 'react'
import React, { useState } from 'react'
import { ExtendComponentsThemeProvider } from '@looker/components-providers'
import { defaultArgTypes as argTypes } from '../../../../../../apps/storybook/src/defaultArgTypes'
import { Button } from '../../../Button'
import type { FieldColorProps } from './FieldColor'
import { FieldColor } from './FieldColor'

export default {
  argTypes,
  component: FieldColor,
  title: 'FieldColor',
}

const Template: Story<
  FieldColorProps & { initialValue: string; externalLabel: boolean }
> = ({ externalLabel = true, initialValue, ...args }) => {
  const [value, setValue] = useState(initialValue)
  const handleChange = (e: FormEvent<HTMLInputElement>) =>
    setValue(e.currentTarget.value)

  return (
    <ExtendComponentsThemeProvider
      themeCustomizations={{ defaults: { externalLabel } }}
    >
      <FieldColor {...args} value={value} onChange={handleChange} />
    </ExtendComponentsThemeProvider>
  )
}

export const Basic = Template.bind({})
Basic.args = { initialValue: 'purple', label: 'Basic' }

export const Required = Template.bind({})
Required.args = {
  ...Basic.args,
  required: true,
}

export const FloatingLabel = Template.bind({})
FloatingLabel.args = { ...Basic.args, externalLabel: false }

export const FloatingLabelNoDefaultValue = Template.bind({})
FloatingLabelNoDefaultValue.args = {
  externalLabel: false,
  initialValue: undefined,
  label: 'Example',
}

export const ControlledFloatingLabelNoValue = () => {
  const [blue, setBlue] = useState<string>('')
  const onClickSetColorToBlue = () => setBlue('blue')
  return (
    <ExtendComponentsThemeProvider
      themeCustomizations={{ defaults: { externalLabel: false } }}
    >
      <Button onClick={onClickSetColorToBlue}>Blue</Button>
      <FieldColor externalLabel={false} label="Controlled" value={blue} />
    </ExtendComponentsThemeProvider>
  )
}
ControlledFloatingLabelNoValue.parameters = {
  storyshots: { disable: true },
}

export const ControlledFloatingLabel = () => {
  const [blue, setBlue] = useState<string>('white')
  const onClickSetColorToBlue = () => setBlue('blue')
  return (
    <ExtendComponentsThemeProvider
      themeCustomizations={{ defaults: { externalLabel: false } }}
    >
      <Button onClick={onClickSetColorToBlue}>Blue</Button>
      <FieldColor externalLabel={false} label="Controlled" value={blue} />
    </ExtendComponentsThemeProvider>
  )
}
ControlledFloatingLabel.parameters = {
  storyshots: { disable: true },
}
