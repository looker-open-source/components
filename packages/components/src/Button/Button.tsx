/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const Button = styled(ButtonBase)`
  background: ${({ theme, color = 'key' }) => theme.colors[color]};
  border: 1px solid ${({ theme, color = 'key' }) => theme.colors[color]};
  color: ${({ theme, color = 'key' }) => theme.colors[`${color}Text`]};

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
  }

  &[aria-expanded='true'],
  &:active,
  &.active {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'key' }) => theme.colors[color]};
      border-color: ${({ theme }) => theme.colors.ui3};
    }
  }
`
