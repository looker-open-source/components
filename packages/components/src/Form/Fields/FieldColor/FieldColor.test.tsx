/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import React from 'react'
import { fireEvent } from '@testing-library/react'
import {
  assertSnapshotShallow,
  renderWithTheme,
} from '@looker/components-test-utils'

import { Button } from '../../../Button'
import { FieldColor } from './FieldColor'

describe('FieldColor', () => {
  test('Default render', () => {
    assertSnapshotShallow(<FieldColor />)
  })

  test('with hidden input', () => {
    const { queryByDisplayValue } = renderWithTheme(
      <FieldColor value="yellow" hideInput />
    )
    expect(queryByDisplayValue('yellow')).not.toBeInTheDocument()
  })

  test('starts with a named color value', () => {
    const { getByDisplayValue } = renderWithTheme(<FieldColor value="green" />)
    expect(getByDisplayValue('green')).toBeInTheDocument()
  })

  test('responds to input value change', () => {
    const { getByDisplayValue } = renderWithTheme(<FieldColor value="green" />)
    const input = getByDisplayValue('green')
    input.focus()
    fireEvent.change(input, { target: { value: 'blue' } })
    expect(getByDisplayValue('blue')).toBeInTheDocument()
  })

  const FieldColorValidationMessage = () => {
    return (
      <FieldColor validationMessage={{ message: 'Error!', type: 'error' }} />
    )
  }

  test('with a validation message', () => {
    const { queryByText } = renderWithTheme(<FieldColorValidationMessage />)
    expect(queryByText('Error!')).toBeInTheDocument()
  })

  test('with an onChange', () => {
    const onChangeMock = jest.fn()
    const { getByLabelText } = renderWithTheme(
      <FieldColor onChange={onChangeMock} label="Background Color" />
    )
    const input = getByLabelText('Background Color')
    fireEvent.change(input, { target: { value: '#FFFF00' } })
    expect(onChangeMock).toHaveBeenCalledWith({
      currentTarget: { value: '#FFFF00' },
      target: { value: '#FFFF00' },
    })
  })

  test('with a defaultValue', () => {
    const { getByLabelText } = renderWithTheme(
      <FieldColor defaultValue="purple" label="Background Color" />
    )
    const input = getByLabelText('Background Color')
    expect(input).toHaveValue('purple')
  })

  test('with controlled state', () => {
    function Wrapper() {
      const [value, setValue] = React.useState('')
      function handleClick() {
        setValue('yellow')
      }
      function handleChange(e: React.FormEvent<HTMLInputElement>) {
        setValue(e.currentTarget.value)
      }
      return (
        <>
          <Button onClick={handleClick}>Turn yellow</Button>
          <FieldColor
            value={value}
            onChange={handleChange}
            placeholder="Select a color"
          />
        </>
      )
    }
    const { getByText, getByPlaceholderText } = renderWithTheme(<Wrapper />)

    const button = getByText('Turn yellow')
    const input = getByPlaceholderText('Select a color')
    expect(input).toHaveValue('')
    fireEvent.click(button)
    expect(input).toHaveValue('yellow')
    fireEvent.change(input, { target: { value: 'purple' } })
    expect(input).toHaveValue('purple')
  })
})
