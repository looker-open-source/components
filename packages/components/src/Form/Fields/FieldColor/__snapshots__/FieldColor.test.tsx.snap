// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldColor Default render 1`] = `ShallowWrapper {}`;
